Add msys2/mingw CI build for the external libraries

adrpo opened the pull request:
This PR adds a CI build for the external libraries on mingw. [It previously just added the  binaries directly].

Partially fixes #29.

tbeu commented:
Thanks for the PR. As libexpat was updated by (8a68ebd9abede31f60a247fb01584082ec7fc390) I'd like to ask you if you could add a Appveyor CI to build the mingw32/64 libs there. Having it as integrated CI would be of greater value than just updating the libs occasionally. Thanks again.

adrpo commented:
Sure. I don't have too much experience with Appveyor CI but I'll see what I can do.

tbeu commented:
I think I can provide teh initial setup. Would be good to have it before end of month such that i can update the MoCo paper. Thanks.

adrpo commented:
I will try using msys2 package as explained here: https://docs.travis-ci.com/user/reference/windows/#how-do-i-use-msys2.
Will update my PR with it soon.

tbeu commented:
Initial setup might be tricky: Have a look at https://github.com/tbeu/matio/blob/master/.appveyor.yml#L133

tbeu commented:
Please rebase your branch on current master and note that libED_SSVFile was replaced by libED_XML2File with new dependency on libxml2. Thanks.

adrpo committed:
WIP
- attempt to add a msys2/mingw64 CI support for building the libraries
- surely not working properly yet, it needs more mingw64 packages installed

adrpo commented:
This needs more work but I will keep working on in.

adrpo committed:
- create the target directory if is not there
- enable just the windows build for now until it works
- put the commands in a file to get proper errors
- add a lot of echo to understand where it goes wrong
- install more packages and both 32bit/64bit ones

adrpo commented:
At least it builds now, I'll clean it up tomorrow or so and then request a review.

overtaker committed:
Upload artifacts

overtaker committed:
Set executable property

tbeu commented on .travis.yml:
```diff
@@ -20,46 +20,112 @@ env:
+        ## nothing yet
+        ;;
+      windows)
+        $mingw64 ./.msys2-build.sh && ./.msys2-build.sh mingw64
```
* Can we please have two configurations in the build matrix?
* Should targetdir always be set?
* Should targetdir be win32 and win64 where binaries go to compiler specific subdir?

adrpo replied:
Done.

adrpo replied:
For mingw we should have mingw64/mingw32 for Visual Studio we should have win32 and win64.

tbeu commented:
@adrpo I am going to release v3.0.0 latest by Friday (30/7) night. Would be good to have your build script CI merged. Thanks.

adrpo commented:
Agreed, will play with it tomorrow.

adrpo committed:
use .build.sh script for all builds, build 32bit and 64bit for mingw

tbeu commented:
Some issues with paths (tar and upload).

tbeu commented:
Can we put the binaries to Library/win32/gcc810 and Library/win64/gcc810 which would be compliant with MLS 3.4?

adrpo commented:
Seems to be OK with the spec: https://specification.modelica.org/maint/3.5/functions.html#annotations-for-external-libraries-and-include-files
I am unsure if we support this layout in OpenModelica but I will check. I know that we support mingw64 and mingw32 at the same level with win32 and win64.

adrpo committed:
fix passing of cmake generator parameter, add macos build call

tbeu commented:
> I am unsure if we support this layout in OpenModelica but I will check. I know that we support mingw64 and mingw32 at the same level with win32 and win64.

I can try with symlinks.

adrpo committed:
fix copy paste error, upload macos builds as well

adrpo committed:
is .build.sh not build.sh

tbeu commented:
Seems, the Travis CI requires more credits. I already asked their support (as I now usually need to do).

adrpo committed:
use the build matrix to set build env variables

adrpo commented:
Hmm, I guess I hit some sort of quota for builds as the last build was cancelled.

adrpo commented:
Ah, OK, i now saw your comment. I guess we can wait with it until tomorrow or so.

tbeu commented:
Yep, just need to wait to see some more credits

![image](https://user-images.githubusercontent.com/5444067/127383715-6338ad1a-e7ae-4598-a1ae-d4a90022b552.png)

adrpo committed:
some comments

adrpo committed:
use bash for popd/pushd

adrpo committed:
export MAKE env variable for each build

adrpo committed:
fix issues with error: unknown type name 'locale_t' on osx

adrpo commented:
Seems the builds are fine now but their upload to bitbucket fails.

```
    [!] error: missing arguments :(
     syntax: upload-to-bitbucket.sh <user> <password> <repo downloads page> <local file to upload>
    example: upload-to-bitbucket.sh swyter secret1 /Swyter/tld-downloads/downloads myfile.zip
    tip: if a file with the same name is already in there it will be overwrited without notice
```

I have no idea how to fix this one.

adrpo commented:
For mingw i actually get:
```
actual upload progress should appear right now as a progress bar, be patient:
[!] server error: bitbucket (the platform) returned HTTP error code 000
```
from the upload script.

tbeu commented:
There is another error:

```
Moving libxml2.a
+mv libxml2.a ../../ExternData/Resources/Library/linux64/libxml2.a
mv: cannot stat 'libxml2.a': No such file or directory
+popd
./.build.sh: 27: ./.build.sh: popd: not found
+echo Moving libxml2.a
Moving libxml2.a
```

tbeu commented:
> Seems the builds are fine now but their upload to bitbucket fails.
> 
> ```
>     [!] error: missing arguments :(
>      syntax: upload-to-bitbucket.sh <user> <password> <repo downloads page> <local file to upload>
>     example: upload-to-bitbucket.sh swyter secret1 /Swyter/tld-downloads/downloads myfile.zip
>     tip: if a file with the same name is already in there it will be overwrited without notice
> ```
> 
> I have no idea how to fix this one.

I can only assume that pathes do not match.

tbeu commented:
Can you please force push your changes again. I did not notice that you commited when I triggered the build again. Sorry for that,

overtaker committed:
Change dir

adrpo commented:
I think I found what the problem is, the build is building in a specific directory and the upload script is called from the top level, will fix it.

adrpo commented:
You already fixed it, I have some other way of fixing it without the cd, basically using the path to the upload file:
`$mingw32 ./upload-to-bitbucket.sh tbeu $BBPASS /tbeu/downloads/downloads ./ExternData/Resources/Library/$PLATFORM/ExternData_$PLATFORM.tar.xz`

Should I push my change or just leave yours?

overtaker committed:
Add after success again

tbeu commented:
Hang on, I am about to fix it.

adrpo commented:
Ok. Let me know if you need anything more.

adrpo commented:
Note that you cannot do:

```
$mingw64 cd someplace
$mingw64 copy from someplace
```
as each command starts from the top again, it has no previous context, so do:
```
$mingw64 cd someplace && copy from someplace
```

tbeu commented:
Can we use cmd or ps1 from https://bitbucket.org/swyter/bitbucket-curl-upload-to-repo-downloads/src/default/ for the uploads as it gives syntax errors otherwise?

tbeu commented:
Here's my current setup: https://github.com/modelica-3rdparty/ExternData/tree/AddMinGWOpenModelicaLibs

overtaker committed:
Test upload

tbeu commented:
Sorry, I cancelled your build as we will run out of credits to soon then.

adrpo commented:
No problem. I will check on my computer if is works fine on my local mingw.

adrpo commented:
With the latest changes it seems to be working fine on my computer. You can try to run it again tomorrow, sorry I had so many tentative changes, is hard to test properly otherwise.

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

tbeu commented:
It still gives "error: missing arguments :(" when uploading. I have no more clue,

overtaker committed:
Test upload again

overtaker committed:
Test upload again

overtaker committed:
Test upload again

tbeu commented:
All merged. Thanks a lot!

tbeu commented:
An this is today's usage statistics:

![image](https://user-images.githubusercontent.com/5444067/127570636-1a3473d9-51c6-4bca-88a5-ac91a16feb91.png)
